src/rules/no-hooks.ts

```diff
@@ -0,0 +1,50 @@
+import { HookName, createRule, isHook } from './tsUtils';
+
+export default createRule({
+  name: __filename,
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Disallow setup and teardown hooks',
+      recommended: false,
+    },
+    messages: {
+      unexpectedHook: "Unexpected '{{ hookName }}' hook",
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          allow: {
+            type: 'array',
+            contains: ['beforeAll', 'beforeEach', 'afterAll', 'afterEach'],
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+    type: 'suggestion',
+  },
+  defaultOptions: [{ allow: [] } as { allow: readonly HookName[] }],
+  create(context, [{ allow }]) {
+    const whitelistedHookNames = allow.reduce((hashMap, value) => {
+      hashMap[value] = true;
+      return hashMap;
+    }, Object.create(null));
+
+    const isWhitelisted = (node: { callee: { name: string } }) =>
```

Comment:
```suggestion
    const isWhitelisted = (node: JestFunctionCallExpressionWithIdentifierCallee<HookName>) =>
```

Reply:
Could we possibly do `export type JestHookFunction = JestFunctionCallExpressionWithIdentifierCallee<HookName>` or something? 🙂

Reply:
> Could we possibly do...

fo sho

>The name is so long... 😛

Yeah, it is - I might do a clean up PR soon, and rename them to something like `Fn` to make them a touch shorter.

Sadly, that's just pretty much what always happens when working w/ AST nodes - there's so much contextual type information that you end up w/ really very verbose but long type names :joy:

Reply:
Yeah, we can go through after everything is migrated. I bet we'll see some patterns or something we can abstract out to make it a bit more readable at that point